src/Test/Perf/SyntaxRewriterBenchmark/Program.cs

```diff
@@ -0,0 +1,458 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+extern alias Baseline;
+using System;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+using static Roslyn.SyntaxRewriterBenchmark.Program.BuiltInFileSets;
+using System.Runtime.Loader;
+
+namespace Roslyn.SyntaxRewriterBenchmark;
+
+public class Program
```

Comment:
While I appreciate the work that went into this benchmark tool, we prefer to use Benchmark.NET, as it will do a lot of work to make sure that the run results are statistically correct, including things like ensuring that JIT time is accounted for. Please remove this project and put the tests into our existing benchmark project, which can be found here: https://github.com/dotnet/performance/tree/main/src/benchmarks/real-world/Roslyn.

Reply:
If Benchmark.NET supports a macro-benchmark paradigm where it can time methods within a long-running task like analyzing the issue test file (which currently takes ~30 seconds), then please share!
Does Roslyn currently have a benchmark that does something similar to this?

Repeating the analysis of a simple, small C# file, and looking at the average time it takes to analyze that file excluding JIT, warmup etc--something that definitely seems suited to Benchmark.NET--would be very different from this syntax rewriter benchmark where analysis of the dataset is expected to take a long time, and where major real-world factors of the test include file I/O, background processes, CPU throttling, input data variability, and everything else that normally goes on within Roslyn-based tasks (like those in Visual Studio), where more detailed metrics (e.g. standard deviation, P90) on how a method performs compared to its baseline can be added as desired (on a file-by-file, node-by-node, and/or token-by-token basis).

Reply:
Bdn should work well for a benchmark that just runs the rewriter on real world files. This is all going to be cpu bound work, so it should show up very well there.

Reply:
I wouldn't expect I/O to be relevant to this benchmark, and it would concern me if it was :slightly_smiling_face:. We're benchmarking the performance of the rewriter, not .NET I/O. We don't have control over optimizing that, and including it adds noise to the metrics we can actually profile for in Roslyn. BDN will do a good job at controlling for those other factors as well, which is why I would like this benchmark there, rather than here, giving us the best opportunity to collect traces (with BDN's profiling option if we choose) and seeing what the change's impacts on the parts we can actually optimize are.

Reply:
I/O is relevant in the sense that the overall gains enabled by this PR are demonstrated by time as a percentage of the total test time, and how improvements perform in real-word scenarios was the question of a team lead prior to me raising this PR...
AFAIK, BDN controls for those factors by repeatedly re-running benchmarks i.e. not a real-world scenario...
@333fred Not sure how long it takes to run the existing real-world Roslyn benchmark that you linked to,
but I guess if we were only interested in the wall-clock time of analyzing the provided set of files,
leveraging BDN would be doable...
But then I'd ask the following questions:
what's the perf of these changes on my machine compared to the baseline?
if LongRunningBenchmark_Baseline is not running at the same time as LongRunningBenchmark_Latest,
how would such a perf comparison be reliable?
how many iterations of such a benchmark would give us the confidence we desire?
how long would such a number of iterations take?
what duration of time would we be okay with? 30 minutes? 1 hour? 2 hours? infinity?

The goal of this PR's syntax rewriter benchmark is to demonstrate real-world improvements to perf over a large, variable dataset, and I really don't see how the same could be done reliably in a reasonable amount of time using BDN,
but further clarity on the above questions would help...
If there isn't an existing benchmark/tool that measures the perf of syntax analysis in a way desired by the team,
perhaps that's a gap to be fulfilled by a separate work item upon which this PR could depend?

Reply:
> Bdn should work well for a benchmark that just runs the rewriter on real world files. This is all going to be cpu bound work, so it should show up very well there

You can just have the test load the text and parse it once. Then you can test the rewriting perf before after this change :-)

Reply:
> what's the perf of these changes on my machine compared to the baseline?

Well, running BDN multiple times, once with the changed version of the reference, once with the original.

> if LongRunningBenchmark_Baseline is not running at the same time as LongRunningBenchmark_Latest,
> how would such a perf comparison be reliable?

The exact same thing could be said for your current benchmarking code: you're not running the changed version of the rewriter at the exact same time as the original version, so what if something changes on your machine between the two runs? Running many times and averaging results is the _only_ way to control for these factors.

> how many iterations of such a benchmark would give us the confidence we desire?

Go with the defaults. I expect this change to be visible on a benchmark that is as simple as `return new CustomSyntaxRewriter().Visit(inputSyntaxNode);` (indeed, that is where it should be _most_ visible).

> The goal of this PR's syntax rewriter benchmark is to demonstrate real-world improvements to perf over a large, variable dataset, and I really don't see how the same could be done reliably in a reasonable amount of time using BDN

The benchmarks of the type introduced by this PR are where BDN really shines. They're exactly the type of thing that it was made to do, and do reliably.

Reply:
1. Without flushing cached nodes, such a benchmark would heavily favor the sibling token retrieval methods in this PR b/c their performance increases even more when all the nodes of a parent are cached.
2. With the auto-calculated BDN number of iterations, the time it would take to run the benchmark for a dataset with a meaningful amount of variance would be extraordinary long,
so my question of _what total benchmarking duration we would be okay with_ is crucial.
3. Running a BDN-based benchmark before the changes and then running it after esp. for a long-running benchmark would require a highly controlled environment not subject to thermal throttling, competing processes etc, which would be difficult for me to obtain currently.
Since the tool randomly switches the order of using latest and baseline with file-by-file or token-by-token granularity there is a far less chance of environmental pollution unevenly affecting the different contexts. 

Sorry if it seems like I'm being overly defensive about keeping the tool in this PR,
I just wanna be open about the thought process behind that tool...

In the beginning I was reliant on VS profiler for comparing various algorithm variations & for trying to iteratively improve perf, but this became a very cumbersome process.
Use of BDN was heavily considered but a CBA of repetition v. variance for comparing the overall perf of syntax analysis against its baseline in a nice amount of time (for dev purposes) inspired this tool with the nice utility of easily seeing the perf of critical methods. Its intent of quickly giving a fair performance comparison between the latest & baseline code was very useful & profiling it to evaluate its timing compared to VS's sampling showed that they were both largely equivalent with the tool slightly giving advantage to the existing baseline, but to me this was not a bad thing.
Comparing perf due to similar variations was more consistent than comparing the perf of outlandishly different ones, but still the tool would give great insight into what benefit changes could provide.

Nevertheless, for drilling down on the statistically rigorous amount of that benefit, a dedicated perf environment w/ minimal background interference (perhaps running both the baseline and the latest single-threaded CPU work on parallel threads/processes) with the BDN-calculated number of iterations and no real concern for how long this benchmark takes would be the way to go in my mind (while either figuring out how to flush cached nodes between iterations of syntax rewriting or iterating both the parsing and syntax rewriting from in-memory constructs).

Reply:
> Without flushing cached nodes, such a benchmark would heavily favor the sibling token retrieval methods in this PR b/c their performance increases even more when all the nodes of a parent are cached.

You can parse a new file and rewrite that file with uncached data.

Reply:
> With the auto-calculated BDN number of iterations, the time it would take to run the benchmark for a dataset with a meaningful amount of variance would be extraordinary long,
so my question of what total benchmarking duration we 

The default iterations should be fine. If not, we can decide what to do at that point. We haven't found any circumstances related to these functions where that would be a problem though.

Reply:
> Running a BDN-based benchmark before the changes and then running it after esp. for a long-running benchmark would require a highly controlled environment not subject to thermal throttling, competing processes etc, which would be difficult for me to obtain currently.

This really isn't the case in practice. Just have two test functions, one which users the prior mechanism and the one which uses the new one. They can be fun at the same time and the results compared.   We can also validate ourselves, running on our own machines. Thanks!

Reply:
> Sorry if it seems like I'm being overly defensive about keeping the tool in this PR,

So far we've found bdn exceptional fur these tasks. We'd prefer to just stick with that absent a clear problem that makes it unsuitable for this problem. So far such a clear problem has not been evident.  Thanks!

Reply:
> 1. Without flushing cached nodes, such a benchmark would heavily favor the sibling token retrieval methods in this PR b/c their performance increases even more when all the nodes of a parent are cached.

Sure, by all means, we should flush the cached nodes. That's fine. Reparsing from the original string in an `IterationSetup` method shouldn't be expensive, particularly if there's no disk I/O involved.

> 2. With the auto-calculated BDN number of iterations, the time it would take to run the benchmark for a dataset with a meaningful amount of variance would be extraordinary long,

This should not be true. A smaller data size than an 18K line file should _absolutely_ reveal perf improvements from this type of change; if it doesn't, that actually concerns me, because this is adding a good amount of complexity to the code. 18K lines is not at all typical for a C# file, and making sure that we don't regress standard C# file perf would be good to do.

> 3. Running a BDN-based benchmark before the changes and then running it after esp. for a long-running benchmark would require a highly controlled environment not subject to thermal throttling, competing processes etc, which would be difficult for me to obtain currently.

We can certainly help with running benchmarks, but I would seriously not expect a benchmark that reveals perf improvements from this type of change to have a serious risk of being subject to these factors.

To put it another way, I understand the motivation of using a macrobenchmark like you have here to make the trouble spots really stand out in a profiler. But I do not think that it is a good idea to check this into the Roslyn codebase. This would be a rarely-run tool that would be subject to bitrot, as it slowly gets out of sync with whatever the latest thing we've done in the compiler is. We already have this problem with the existing compiler benchmarks, and this would be worse, since we'd run it even less often. I also don't want to spend review time on the tool, and trying to determine if the results are statistically significant, especially when we already have an excellent microbenchmarking tool that will nicely show the impact of this PR on files of various sizes.

Reply:
@333fred Tool removed; I've presented my analysis of this PR's perf improvements so outside analysis (more likely to be trusted & accepted) would be awesome!